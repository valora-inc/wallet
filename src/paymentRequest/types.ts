import { NotificationTypes } from 'src/notifications/types'

export enum PaymentRequestStatus {
  REQUESTED = 'REQUESTED',
  COMPLETED = 'COMPLETED',
  DECLINED = 'DECLINED',
  CANCELLED = 'CANCELLED',
}

export interface PaymentRequest {
  uid?: string
  amount: string
  comment?: string
  createdAt: number
  requesterAddress: string
  requesteeAddress: string
  requesterE164Number?: string
  status: PaymentRequestStatus
  notified: boolean
  type?: NotificationTypes.PAYMENT_REQUESTED
}

export interface WriteablePaymentRequest extends Omit<PaymentRequest, 'createdAt'> {
  createdAt: object
}
