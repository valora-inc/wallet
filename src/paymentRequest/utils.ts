import { decryptComment, encryptComment } from '@celo/cryptographic-utils'
import { isE164NumberStrict } from '@celo/phone-utils'
import { hexToBuffer } from '@celo/utils/lib/address'
import BigNumber from 'bignumber.js'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import { Recipient } from 'src/recipients/recipient'
import { TransactionDataInput } from 'src/send/SendAmount'
import { TokenBalance } from 'src/tokens/slice'
import Logger from 'src/utils/Logger'
import { Currency } from 'src/utils/currencies'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Convert a payment request into a transaction data object if user has enough balance
export const transactionDataFromPaymentRequest = ({
  paymentRequest,
  stableTokens,
  requester,
}: {
  paymentRequest: PaymentRequest
  stableTokens: TokenBalance[]
  requester: Recipient
}): TransactionDataInput => {
  const cUsdTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Dollar)
  const cEurTokenInfo = stableTokens.find((token) => token?.symbol === Currency.Euro)
  if (!cUsdTokenInfo?.address || !cEurTokenInfo?.address) {
    // Should never happen in production
    throw new Error('No token address found for cUSD or cEUR')
  }
  // If the user has enough cUSD balance, pay with cUSD
  // Else, try with cEUR
  // Else, throw error
  const usdRequested = new BigNumber(paymentRequest.amount)

  if (
    cUsdTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cUsdTokenInfo.balance.multipliedBy(cUsdTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount),
      tokenAmount: new BigNumber(paymentRequest.amount),
      amountIsInLocalCurrency: false,
      tokenAddress: cUsdTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  } else if (
    cEurTokenInfo.usdPrice &&
    usdRequested.isLessThanOrEqualTo(cEurTokenInfo.balance.multipliedBy(cEurTokenInfo.usdPrice))
  ) {
    return {
      comment: paymentRequest.comment,
      recipient: requester,
      inputAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      tokenAmount: new BigNumber(paymentRequest.amount).dividedBy(cEurTokenInfo.usdPrice),
      amountIsInLocalCurrency: false,
      tokenAddress: cEurTokenInfo.address,
      paymentRequestId: paymentRequest.uid || '',
    }
  }
  throw new Error('Insufficient stablecoin balance')
}

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164NumberStrict(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable') ?? undefined
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
