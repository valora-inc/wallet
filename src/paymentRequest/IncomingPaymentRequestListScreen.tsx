import React from 'react'
import { WithTranslation } from 'react-i18next'
import { View } from 'react-native'
import { connect } from 'react-redux'
import i18n, { withTranslation } from 'src/i18n'
import { HeaderTitleWithBalance } from 'src/navigator/Headers'
import { NotificationList } from 'src/notifications/NotificationList'
import IncomingPaymentRequestListItem from 'src/paymentRequest/IncomingPaymentRequestListItem'
import { getIncomingPaymentRequests } from 'src/paymentRequest/selectors'
import { PaymentRequest } from 'src/paymentRequest/types'
import { getRecipientFromAddress, RecipientInfo } from 'src/recipients/recipient'
import { recipientInfoSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'
import { Currency } from 'src/utils/currencies'

interface StateProps {
  paymentRequests: PaymentRequest[]
  recipientInfo: RecipientInfo
}

const mapStateToProps = (state: RootState): StateProps => ({
  paymentRequests: getIncomingPaymentRequests(state),
  recipientInfo: recipientInfoSelector(state),
})

type Props = WithTranslation & StateProps

export const listItemRenderer = (props: { recipientInfo: RecipientInfo }) => (
  request: PaymentRequest,
  key: number | undefined = undefined
) => (
  <View key={key}>
    <IncomingPaymentRequestListItem
      id={request.uid || ''}
      amount={request.amount}
      requester={getRecipientFromAddress(request.requesterAddress, props.recipientInfo)}
      comment={request.comment}
    />
  </View>
)

class IncomingPaymentRequestListScreen extends React.Component<Props> {
  static navigationOptions = () => ({
    headerTitle: (
      <HeaderTitleWithBalance title={i18n.t('incomingPaymentRequests')} token={Currency.Dollar} />
    ),
  })

  render = () => {
    return (
      <NotificationList
        items={this.props.paymentRequests}
        listItemRenderer={listItemRenderer(this.props)}
      />
    )
  }
}

export default connect<StateProps, {}, {}, RootState>(
  mapStateToProps,
  {}
)(withTranslation<Props>()(IncomingPaymentRequestListScreen))
