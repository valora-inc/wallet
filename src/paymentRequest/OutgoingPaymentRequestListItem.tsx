import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import RequestMessagingCard from 'src/components/RequestMessagingCard'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { withTranslation } from 'src/i18n'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import { getDisplayName, Recipient } from 'src/recipients/recipient'
import { Currency } from 'src/utils/currencies'
import Logger from 'src/utils/Logger'

interface OwnProps {
  id: string
  requestee: Recipient
  amount: string
  comment?: string
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
  index?: number
}

type Props = OwnProps & WithTranslation

export class OutgoingPaymentRequestListItem extends React.Component<Props> {
  onRemind = () => {
    const { id, t, index } = this.props
    this.props.updatePaymentRequestNotified(id, false)
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.remind,
      notificationPositionInList: index,
    })
    Logger.showMessage(t('reminderSent'))
  }

  onCancel = () => {
    const { id, index } = this.props
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.decline,
      notificationPositionInList: index,
    })
    this.props.cancelPaymentRequest(id)
  }

  getCTA = () => {
    return [
      {
        text: this.props.t('remind'),
        onPress: this.onRemind,
      },
      {
        text: this.props.t('cancel'),
        onPress: this.onCancel,
      },
    ]
  }

  render() {
    const { requestee, id, comment, t } = this.props
    const amount = {
      value: this.props.amount,
      currencyCode: Currency.Dollar,
    }

    return (
      <RequestMessagingCard
        testID={`OutgoingPaymentRequestNotification/${id}`}
        title={t('outgoingPaymentRequestNotificationTitle', {
          name: getDisplayName(requestee, t),
        })}
        amount={<CurrencyDisplay amount={amount} />}
        details={comment}
        icon={<ContactCircle recipient={requestee} />}
        callToActions={this.getCTA()}
      />
    )
  }
}

export default withTranslation<Props>()(OutgoingPaymentRequestListItem)
