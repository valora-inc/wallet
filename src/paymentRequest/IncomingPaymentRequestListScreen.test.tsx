import { render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import IncomingPaymentRequestListScreen from 'src/paymentRequest/IncomingPaymentRequestListScreen'
import { PaymentRequest } from 'src/paymentRequest/types'
import { createMockPaymentRequest } from 'src/paymentRequest/testValues'
import { Currency } from 'src/utils/currencies'
import { createMockStore } from 'test/utils'
import { mockAccount, mockE164Number } from 'test/values'

const requests = [
  createMockPaymentRequest({
    amount: '20',
    comment: 'Just the best',
    requesterE164Number: '+1555-867-5309',
    requesterAddress: mockAccount,
  }),
  createMockPaymentRequest({
    amount: '102',
    comment: 'Just the best for the best. Thanos & Zeus Gods of ultimate Power',
    requesterE164Number: mockE164Number,
    requesterAddress: mockAccount,
  }),
  createMockPaymentRequest({
    amount: '1',
    comment: 'Just the best but less',
    requesterE164Number: mockE164Number,
    requesterAddress: mockAccount,
  }),
]

function testStore(incomingPaymentRequests: PaymentRequest[]) {
  return createMockStore({
    stableToken: { balances: { [Currency.Dollar]: '120', [Currency.Euro]: '10' } },
    paymentRequest: { incomingPaymentRequests },
  })
}

describe('IncomingPaymentRequestListScreen', () => {
  it('renders correctly with requests', () => {
    const store = testStore(requests)

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListScreen />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly with no requests', () => {
    const store = testStore([])

    const tree = render(
      <Provider store={store}>
        <IncomingPaymentRequestListScreen />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
