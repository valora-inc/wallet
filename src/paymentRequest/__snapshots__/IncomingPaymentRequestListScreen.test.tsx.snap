// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomingPaymentRequestListScreen renders correctly with no requests 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#2E3338",
          "fontFamily": "Inter-Regular",
          "fontSize": 16,
          "lineHeight": 22,
        },
        {
          "marginTop": 30,
          "textAlign": "center",
        },
      ]
    }
  >
    emptyList
  </Text>
</RNCSafeAreaView>
`;

exports[`IncomingPaymentRequestListScreen renders correctly with requests 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "margin": 16,
          }
        }
      >
        <View
          style={
            {
              "marginBottom": 16,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "#FFFFFF",
                  "padding": 16,
                },
                {
                  "borderRadius": 8,
                },
                undefined,
                [
                  {
                    "borderColor": "#E6E6E6",
                    "borderWidth": 1,
                    "flex": 1,
                    "minHeight": 144,
                  },
                  {},
                ],
              ]
            }
            testID="IncomingPaymentRequestNotification/7363bvby2ba8270273"
          >
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "marginRight": 16,
                  }
                }
              >
                <Text
                  numberOfLines={2}
                  style={
                    {
                      "color": "#666666",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 13,
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Title"
                >
                  incomingPaymentRequestNotificationTitle, {"name":"John Doe"}
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-SemiBold",
                      "fontSize": 24,
                      "lineHeight": 28,
                      "marginBottom": 2,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Amount"
                >
                  <Text
                    style={
                      [
                        undefined,
                        {
                          "color": undefined,
                        },
                      ]
                    }
                    testID="undefined/value"
                  >
                    â‚±
                    26.60
                  </Text>
                </Text>
                <Text
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Details"
                >
                  Just the best
                </Text>
              </View>
              <View
                style={{}}
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        {
                          "backgroundColor": "hsl(0, 53%, 93%)",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        [
                          {
                            "color": "#FFFFFF",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 16,
                          },
                          {
                            "color": "hsl(0, 67%, 24%)",
                            "fontSize": 20,
                          },
                        ]
                      }
                    >
                      J
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "flex-start",
                }
              }
              testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions"
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/send/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  send
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/decline/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  decline
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "marginBottom": 16,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "#FFFFFF",
                  "padding": 16,
                },
                {
                  "borderRadius": 8,
                },
                undefined,
                [
                  {
                    "borderColor": "#E6E6E6",
                    "borderWidth": 1,
                    "flex": 1,
                    "minHeight": 144,
                  },
                  {},
                ],
              ]
            }
            testID="IncomingPaymentRequestNotification/7363bvby2ba8270273"
          >
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "marginRight": 16,
                  }
                }
              >
                <Text
                  numberOfLines={2}
                  style={
                    {
                      "color": "#666666",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 13,
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Title"
                >
                  incomingPaymentRequestNotificationTitle, {"name":"John Doe"}
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-SemiBold",
                      "fontSize": 24,
                      "lineHeight": 28,
                      "marginBottom": 2,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Amount"
                >
                  <Text
                    style={
                      [
                        undefined,
                        {
                          "color": undefined,
                        },
                      ]
                    }
                    testID="undefined/value"
                  >
                    â‚±
                    135.66
                  </Text>
                </Text>
                <Text
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Details"
                >
                  Just the best for the best. Thanos & Zeus Gods of ultimate Power
                </Text>
              </View>
              <View
                style={{}}
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        {
                          "backgroundColor": "hsl(0, 53%, 93%)",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        [
                          {
                            "color": "#FFFFFF",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 16,
                          },
                          {
                            "color": "hsl(0, 67%, 24%)",
                            "fontSize": 20,
                          },
                        ]
                      }
                    >
                      J
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "flex-start",
                }
              }
              testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions"
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/send/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  send
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/decline/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  decline
                </Text>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "marginBottom": 16,
            }
          }
        >
          <View
            style={
              [
                {
                  "backgroundColor": "#FFFFFF",
                  "padding": 16,
                },
                {
                  "borderRadius": 8,
                },
                undefined,
                [
                  {
                    "borderColor": "#E6E6E6",
                    "borderWidth": 1,
                    "flex": 1,
                    "minHeight": 144,
                  },
                  {},
                ],
              ]
            }
            testID="IncomingPaymentRequestNotification/7363bvby2ba8270273"
          >
            <View
              style={
                {
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  {
                    "flex": 1,
                    "marginRight": 16,
                  }
                }
              >
                <Text
                  numberOfLines={2}
                  style={
                    {
                      "color": "#666666",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 13,
                      "lineHeight": 16,
                      "marginBottom": 4,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Title"
                >
                  incomingPaymentRequestNotificationTitle, {"name":"John Doe"}
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-SemiBold",
                      "fontSize": 24,
                      "lineHeight": 28,
                      "marginBottom": 2,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Amount"
                >
                  <Text
                    style={
                      [
                        undefined,
                        {
                          "color": undefined,
                        },
                      ]
                    }
                    testID="undefined/value"
                  >
                    â‚±
                    1.33
                  </Text>
                </Text>
                <Text
                  style={
                    {
                      "color": "#2E3338",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 14,
                      "lineHeight": 18,
                    }
                  }
                  testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/Details"
                >
                  Just the best but less
                </Text>
              </View>
              <View
                style={{}}
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "justifyContent": "center",
                        },
                        {
                          "backgroundColor": "hsl(0, 53%, 93%)",
                          "borderRadius": 20,
                          "height": 40,
                          "width": 40,
                        },
                      ]
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        [
                          {
                            "color": "#FFFFFF",
                            "fontFamily": "Inter-Medium",
                            "fontSize": 16,
                          },
                          {
                            "color": "hsl(0, 67%, 24%)",
                            "fontSize": 20,
                          },
                        ]
                      }
                    >
                      J
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                {
                  "flexDirection": "row",
                  "flexWrap": "wrap",
                  "justifyContent": "flex-start",
                }
              }
              testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions"
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/send/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  send
                </Text>
              </View>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                focusable={true}
                nativeBackgroundAndroid={
                  {
                    "attribute": "selectableItemBackgroundBorderless",
                    "rippleRadius": undefined,
                    "type": "ThemeAttrAndroid",
                  }
                }
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                testID="IncomingPaymentRequestNotification/7363bvby2ba8270273/CallToActions/decline/Button"
              >
                <Text
                  style={
                    [
                      {
                        "color": "#1AB775",
                        "fontFamily": "Inter-SemiBold",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      {
                        "fontSize": 14,
                        "lineHeight": 16,
                        "marginRight": 24,
                        "minHeight": 16,
                        "minWidth": 48,
                      },
                    ]
                  }
                >
                  decline
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
